### PR TITLE
Fixed https://github.com/aws/amazon-ecs-agent/issues/506

### PR DESCRIPTION
**Pull request by liwenwu-amazon**
### Summary

Fix issue #506 

### Implementation details

1 . Add a new field 'lastResetTime' to stats Queue struct, Set 'lastResetTime' whenever Reseting the Queue.
2. With 1,  the caller can return an error if queue has NOT been filled with expected data points after a resetThreshold period
3. define a new error code engine.EmptyMetricsError, so that the caller can ignore this error when publishing metrics immediately after agent connect to backend. And  return/handle this error in other situations. 

### Testing

- [ x] Builds on Linux (`make release`)
- [ ] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [ x] Unit tests on Linux (`make test`) pass
- [ ] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [ x] Integration tests on Linux (`make run-integ-tests`) pass
- [ ] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x ] Functional tests on Linux (`make run-functional-tests`) pass
- [ ] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: 

### Licensing

This contribution is under the terms of the Apache 2.0 License: 

@richardpen @samuelkarp @aaithal

### CODE REVIEW

File: agent/stats/engine.go [hunk 1]
```diff
@@ -35,6 +36,7 @@ import (
 const (
 	containerChangeHandler = "DockerStatsEngineDockerEventsHandler"
 	listContainersTimeout  = 10 * time.Minute
+	resetThreshold         = 2 * ecsengine.StatsInactivityTimeout
```
Comment (aaithal) [inline agent/stats/engine.go]: Can you rename this to be more descriptive? `queueResetThreshold` perhaps?

File: agent/stats/engine.go [hunk 2]
```diff
@@ -398,6 +401,13 @@ func (engine *DockerStatsEngine) getContainerMetricsForTask(taskArn string) ([]*
 			engine.doRemoveContainer(container, taskArn)
 			continue
 		}
+
+		if !container.statsQueue.resetThreshholdElapsed(resetThreshold) &&
```
Comment (aaithal) [inline agent/stats/engine.go]: Typo: `resetThreshholdElapsed` -> `resetThresholdElapsed`

File: agent/stats/queue.go [hunk 3]
```diff
@@ -133,6 +137,19 @@ func getMemoryUsagePerc(s *UsageStats) float64 {
 
 type getUsageFunc func(*UsageStats) float64
 
+func (queue *Queue) resetThreshholdElapsed(timeOut time.Duration) bool {
```
Comment (aaithal) [inline agent/stats/queue.go]: this should be `timeout` and not `timeOut`.

File: agent/stats/queue_test.go [hunk 4]
```diff
@@ -309,3 +309,64 @@ func TestCpuStatsSetNotSetToInfinity(t *testing.T) {
...
 }
+
+func TestResetThreshholdElapsed(t *testing.T) {
+	// create a queue
+	queueLength := 3
+	queue := NewQueue(queueLength)
+
+	queue.Reset()
+
+	threshholdElapsed := queue.resetThreshholdElapsed(2 * time.Millisecond)
```
Comment (aaithal) [inline agent/stats/queue_test.go]: typo: `threshholdElapsed` -> `thresholdElapsed`

File: agent/stats/queue_test.go [hunk 5]
```diff
@@ -309,3 +309,64 @@ func TestCpuStatsSetNotSetToInfinity(t *testing.T) {
...
+	if threshholdElapsed {
```
Comment (aaithal) [inline agent/stats/queue_test.go]: Can you use `assert.False()` instead of this?

File: agent/stats/queue_test.go [hunk 6]
```diff
@@ -309,3 +309,64 @@ func TestCpuStatsSetNotSetToInfinity(t *testing.T) {
...
+		t.Errorf("Queue is NOT expected to elapse the threshhold right after reset")
+	}
+
+	time.Sleep(3 * time.Millisecond)
+	threshholdElapsed = queue.resetThreshholdElapsed(2 * time.Millisecond)
+
+	if !threshholdElapsed {
```
Comment (aaithal) [inline agent/stats/queue_test.go]: Same here. Please use the `github.com/stretchr/testify/assert` package's `True()` method here as well.

File: agent/stats/queue_test.go [hunk 7]
```diff
@@ -309,3 +309,64 @@ func TestCpuStatsSetNotSetToInfinity(t *testing.T) {
...
+		3649536,
+		3649536,
+		3649536,
+	}
+	// create a queue
+	queueLength := 3
+	queue := NewQueue(queueLength)
+
+	enoughDataPoints := queue.enoughDatapointsInBuffer()
+	if enoughDataPoints {
```
Comment (aaithal) [inline agent/stats/queue_test.go]: Please use `assert.False()` here.

File: agent/stats/queue_test.go [hunk 8]
```diff
@@ -309,3 +309,64 @@ func TestCpuStatsSetNotSetToInfinity(t *testing.T) {
...
+		t.Errorf(" Queue is expected NOT having enough data points right after creation")
+	}
+	for i, time := range timestamps {
+		queue.Add(&ContainerStats{cpuUsage: cpuTimes[i], memoryUsage: memoryUtilizationInBytes[i], timestamp: time})
+	}
+
+	enoughDataPoints = queue.enoughDatapointsInBuffer()
+	if !enoughDataPoints {
```
Comment (aaithal) [inline agent/stats/queue_test.go]: Please use `assert.True()` here.

File: agent/tcs/client/client.go [hunk 9]
```diff
@@ -164,6 +170,7 @@ func (cs *clientServer) publishMetricsOnce() {
 			seelog.Warnf("Error publishing metrics: %v. Request: %v", err, request)
```
Comment (aaithal) [inline agent/tcs/client/client.go]: Shouldn't this return the error immediately as well?

File: agent/stats/engine.go (hunk 1 above)
Comment (liwenwu-amazon) [inline agent/stats/engine.go]: done

File: agent/stats/engine.go (hunk 2 above)
Comment (liwenwu-amazon) [inline agent/stats/engine.go]: done

File: agent/stats/queue.go (hunk 3 above)
Comment (liwenwu-amazon) [inline agent/stats/queue.go]: done

File: agent/stats/queue_test.go (hunk 4 above)
Comment (liwenwu-amazon) [inline agent/stats/queue_test.go]: done

File: agent/stats/queue_test.go (hunk 5 above)
Comment (liwenwu-amazon) [inline agent/stats/queue_test.go]: done

File: agent/stats/queue_test.go (hunk 6 above)
Comment (liwenwu-amazon) [inline agent/stats/queue_test.go]: done

File: agent/stats/queue_test.go (hunk 7 above)
Comment (liwenwu-amazon) [inline agent/stats/queue_test.go]: done

File: agent/stats/queue_test.go (hunk 8 above)
Comment (liwenwu-amazon) [inline agent/stats/queue_test.go]: done

File: agent/tcs/client/client.go [hunk 10]
```diff
@@ -137,33 +137,40 @@ func (cs *clientServer) publishMetrics() {
 	// Publish metrics immediately after we connect and wait for ticks. This makes
 	// sure that there is no data loss when a scheduled metrics publishing fails
 	// due to a connection reset.
-	cs.publishMetricsOnce()
+	err := cs.publishMetricsOnce()
+	if err != nil && err != stats.EmptyMetricsError {
+		seelog.Warnf("Error getting instance metrics: %v", err)
```
Comment (aaithal) [inline agent/tcs/client/client.go]: Can you rephrase the error message? The error is from the the `publishMetricsOnce()` method. We can't be sure here if its because of publish failing or if there was an error getting metrics.

File: agent/tcs/client/client.go [hunk 11]
```diff
@@ -137,33 +137,40 @@ func (cs *clientServer) publishMetrics() {
...
+	}
 	// don't simply range over the ticker since its channel doesn't ever get closed
 	for {
 		select {
 		case <-cs.publishTicker.C:
-			cs.publishMetricsOnce()
+			err := cs.publishMetricsOnce()
+			if err != nil {
+				seelog.Warnf("Error getting instance metrics: %v", err)
```
Comment (aaithal) [inline agent/tcs/client/client.go]: Same as above, can you rephrase the error message?

File: agent/tcs/client/client.go [hunk 12]
```diff
@@ -137,33 +137,40 @@ func (cs *clientServer) publishMetrics() {
...
+			}
 		case <-cs.endPublish:
 			return
 		}
 	}
 }
 
 // publishMetricsOnce is invoked by the ticker to periodically publish metrics to backend.
-func (cs *clientServer) publishMetricsOnce() {
+func (cs *clientServer) publishMetricsOnce() error {
```
Comment (aaithal) [inline agent/tcs/client/client.go]: Can you add/update unit tests to cover this change?

Comment (aaithal) [inline agent/tcs/client/client.go]: Also, the [functional tests for windows](https://github.com/liwenwu-amazon/amazon-ecs-agent/commits/issue506) seem to have just stuck. Can you address that and make sure those pass as well?

File: agent/stats/engine.go [hunk 13]
```diff
@@ -398,6 +401,13 @@ func (engine *DockerStatsEngine) getContainerMetricsForTask(taskArn string) ([]*
 			engine.doRemoveContainer(container, taskArn)
 			continue
 		}
+
+		if !container.statsQueue.resetThresholdElapsed(queueResetThreshold) &&
```
Comment (samuelkarp) [inline agent/stats/engine.go]: I find this conditional pretty hard to read.  I think it might be easier like this:

```go
if !container.statsQueue.enoughDatapointsInBuffer() &&
	!container.statsQueue.resetThresholdElapsed(queueResetThreshold)
```

That lets it read more logically as "if there are not enough datapoints and the threshold is not elapsed".

File: agent/tcs/client/client_test.go [hunk 14]
```diff
@@ -152,6 +158,18 @@ func TestPublishMetricsRequest(t *testing.T) {
 
 	cs.Close()
 }
+func TestPublishMetricsOnceEmptyStatsError(t *testing.T) {
+
```
Comment (samuelkarp) [inline agent/tcs/client/client_test.go]: stray empty line

File: agent/acs/handler/acs_handler_test.go [hunk 15]
```diff
@@ -519,7 +519,7 @@ func TestHandlerReconnectDelayForInactiveInstanceError(t *testing.T) {
 		mockWsClient.EXPECT().Connect().Do(func() {
 			reconnectDelay := time.Now().Sub(firstConnectionAttemptTime)
 			t.Logf("Delay between successive connections: %v", reconnectDelay)
-			if reconnectDelay < inactiveInstanceReconnectDelay {
+			if reconnectDelay+1 < inactiveInstanceReconnectDelay {
```
Comment (liwenwu-amazon) [inline agent/acs/handler/acs_handler_test.go]: On Window platform, we find intermittently that time.Now().Sub(...) can returns 199.9989ms  even after code waiting for time.NewTimer(200) ms

Comment (samuelkarp) [inline agent/acs/handler/acs_handler_test.go]: Rather than blindly adding 1, can you:
* Include this information about intermittent time subtraction issues in an explanatory comment
* Explicitly allow for some slop in the assertion here (possibly up to `1ms`?) by adjusting the expected time rather than the observed time and having a named value instead of a literal (use a constant or variable to describe the modifier).

File: agent/acs/handler/acs_handler_test.go [hunk 16]
```diff
@@ -519,8 +519,10 @@ func TestHandlerReconnectDelayForInactiveInstanceError(t *testing.T) {
 		mockWsClient.EXPECT().Connect().Do(func() {
 			reconnectDelay := time.Now().Sub(firstConnectionAttemptTime)
 			t.Logf("Delay between successive connections: %v", reconnectDelay)
-			if reconnectDelay < inactiveInstanceReconnectDelay {
-				t.Errorf("Reconnect delay %v is not less than inactive instance reconnect delay %v", reconnectDelay, inactiveInstanceReconnectDelay)
+			// On window platform, we found issue with time.Now().Sub(...) reporting 199.9989 even after the code has already waited for time.NewTimer(200)ms.
+			timeSubFuncSlopAllowed := 1 * time.Millisecond
+			if reconnectDelay+timeSubFuncSlopAllowed < inactiveInstanceReconnectDelay {
```
Comment (samuelkarp) [inline agent/acs/handler/acs_handler_test.go]: Can you change this to `reconnectDelay < inactiveInstanceReconnectDelay - timeSubFuncSlopAllowed`?

File: agent/tcs/client/client_test.go [hunk 17]
```diff
@@ -150,6 +156,17 @@ func TestPublishMetricsRequest(t *testing.T) {
 
 	cs.Close()
 }
+func TestPublishMetricsOnceEmptyStatsError(t *testing.T) {
+	cs := clientServer{
+		statsEngine: &emptyStatsEngine{},
+	}
+	err := cs.publishMetricsOnce()
+
+	if err == nil {
```
Comment (samuelkarp) [inline agent/tcs/client/client_test.go]: `assert.NoError(err)`

File: agent/tcs/client/client_test.go [hunk 18]
```diff
@@ -150,6 +156,17 @@ func TestPublishMetricsRequest(t *testing.T) {
...
+		t.Error("Failed: expecting publishMerticOnce return err ")
```
Comment (samuelkarp) [inline agent/tcs/client/client_test.go]: s/Mertic/Metric/

File: agent/tcs/client/client.go (hunk 12 above)
Comment (liwenwu-amazon) [inline agent/tcs/client/client.go]: Opened issue #663 and #662 to track unit-test failure on windows platform

File: agent/acs/handler/acs_handler_test.go [hunk 19]
```diff
@@ -519,8 +519,10 @@ func TestHandlerReconnectDelayForInactiveInstanceError(t *testing.T) {
 		mockWsClient.EXPECT().Connect().Do(func() {
 			reconnectDelay := time.Now().Sub(firstConnectionAttemptTime)
 			t.Logf("Delay between successive connections: %v", reconnectDelay)
-			if reconnectDelay < inactiveInstanceReconnectDelay {
-				t.Errorf("Reconnect delay %v is not less than inactive instance reconnect delay %v", reconnectDelay, inactiveInstanceReconnectDelay)
+			// On window platform, we found issue with time.Now().Sub(...) reporting 199.9989 even after the code has already waited for time.NewTimer(200)ms.
+			timeSubFuncSlopAllowed := 1 * time.Millisecond
```
Comment (aaithal) [inline agent/acs/handler/acs_handler_test.go]: I'm not a fan of doing this math here. Can you please use `assert.WithinDuration` instead? https://godoc.org/github.com/stretchr/testify/assert#Assertions.WithinDuration

Writing less code is always better :)

File: agent/stats/queue_test.go [hunk 20]
```diff
@@ -309,3 +310,55 @@ func TestCpuStatsSetNotSetToInfinity(t *testing.T) {
...
+	thresholdElapsed := queue.resetThresholdElapsed(2 * time.Millisecond)
+	assert.False(t, thresholdElapsed, "Queue is NOT expected to elapsed the threshold right after reset")
+
+	time.Sleep(3 * time.Millisecond)
+	thresholdElapsed = queue.resetThresholdElapsed(2 * time.Millisecond)
+
+	assert.True(t, thresholdElapsed, "Queue is expected to elapse the threshold")
+
+}
+func TestEnoughDatapointsInBuffer(t *testing.T) {
```
Comment (aaithal) [inline agent/stats/queue_test.go]: nit: new line needed

File: agent/acs/handler/acs_handler_test.go [hunk 21]
```diff
@@ -518,11 +518,14 @@ func TestHandlerReconnectDelayForInactiveInstanceError(t *testing.T) {
...
 		mockWsClient.EXPECT().Connect().Do(func() {
 			reconnectDelay := time.Now().Sub(firstConnectionAttemptTime)
+			reconnectDelayTime := time.Now()
 			t.Logf("Delay between successive connections: %v", reconnectDelay)
-			// On window platform, we found issue with time.Now().Sub(...) reporting 199.9989 even after the code has already waited for time.NewTimer(200)ms.
 			timeSubFuncSlopAllowed := 1 * time.Millisecond
-			if reconnectDelay < inactiveInstanceReconnectDelay-timeSubFuncSlopAllowed {
-				t.Errorf("Reconnect delay %v is 1 ms less than inactive instance reconnect delay %v", reconnectDelay, inactiveInstanceReconnectDelay)
+			// On window platform, we found issue with time.Now().Sub(...) reporting 199.9989 even after the code has already waited for time.NewTimer(200)ms.
+			if reconnectDelay < inactiveInstanceReconnectDelay {
```
Comment (aaithal) [inline agent/acs/handler/acs_handler_test.go]: I don't think you need this if block when using `assert.WithinDuration`

File: agent/stats/engine.go [hunk 22]
```diff
@@ -398,6 +401,13 @@ func (engine *DockerStatsEngine) getContainerMetricsForTask(taskArn string) ([]*
 			engine.doRemoveContainer(container, taskArn)
 			continue
 		}
+
+		if !container.statsQueue.enoughDatapointsInBuffer() &&
+			!container.statsQueue.resetThresholdElapsed(queueResetThreshold) {
+			seelog.Debug("Stats not ready for container %s", dockerID)
```
Comment (aaithal) [inline agent/stats/engine.go]: `Debugf`

File: agent/acs/handler/acs_handler_test.go (hunk 21 above)
Comment (liwenwu-amazon) [inline agent/acs/handler/acs_handler_test.go]: There are 2 conditions here
1) to make sure reconnectDelay is "greater" than inactiveInstanceReconnectDelay
2) or to make sure it is within duration of the 1 milisecond, if reconnectDelay is "less" than inactiveinstanceReconnectDelay
The code here is intended for "1"

Comment (aaithal) [inline agent/acs/handler/acs_handler_test.go]: I don't think that's entirely correct. I think the intention is to ensure that we wait for `inactiveInstanceReconnectDelay` between disconnects. It's `>= inactiveInstanceReconnectDelay`. We don't add any jitter when the disconnection is for an inactive instance. So, `assert.WithinDuration()` should be good enough with a buffer of `2ms`.

File: agent/acs/handler/acs_handler_test.go [hunk 23]
```diff
@@ -518,10 +518,13 @@ func TestHandlerReconnectDelayForInactiveInstanceError(t *testing.T) {
 		}).Return(fmt.Errorf("InactiveInstanceException:")),
 		mockWsClient.EXPECT().Connect().Do(func() {
 			reconnectDelay := time.Now().Sub(firstConnectionAttemptTime)
+			reconnectDelayTime := time.Now()
 			t.Logf("Delay between successive connections: %v", reconnectDelay)
-			if reconnectDelay < inactiveInstanceReconnectDelay {
-				t.Errorf("Reconnect delay %v is not less than inactive instance reconnect delay %v", reconnectDelay, inactiveInstanceReconnectDelay)
-			}
+			timeSubFuncSlopAllowed := 2 * time.Millisecond
+			// On window platform, we found issue with time.Now().Sub(...) reporting 199.9989 even after the code has already waited for time.NewTimer(200)ms.
```
Comment (aaithal) [inline agent/acs/handler/acs_handler_test.go]: As @samuelkarp mentioned in one of the earlier comments, there's no "window platform", please modify that. Also, please split the comment across multiple lines to reduce the line width.

File: agent/stats/queue_test.go [hunk 24]
```diff
@@ -309,3 +310,55 @@ func TestCpuStatsSetNotSetToInfinity(t *testing.T) {
...
+
+func TestResetThresholdElapsed(t *testing.T) {
+	// create a queue
+	queueLength := 3
+	queue := NewQueue(queueLength)
+
+	queue.Reset()
+
+	thresholdElapsed := queue.resetThresholdElapsed(2 * time.Millisecond)
+	assert.False(t, thresholdElapsed, "Queue is NOT expected to elapsed the threshold right after reset")
```
Comment (aaithal) [inline agent/stats/queue_test.go]: I don't think the phrasing is correct here. The Queue does not elapse any threshold as it's not an entity that is time-bound. I think the better message here is "Queue reset threshold is not expected to elapse right after reset"

File: agent/stats/queue_test.go [hunk 25]
```diff
@@ -309,3 +310,55 @@ func TestCpuStatsSetNotSetToInfinity(t *testing.T) {
...
+
+	time.Sleep(3 * time.Millisecond)
+	thresholdElapsed = queue.resetThresholdElapsed(2 * time.Millisecond)
+
+	assert.True(t, thresholdElapsed, "Queue is expected to elapse the threshold")
```
Comment (aaithal) [inline agent/stats/queue_test.go]: Same as above. I think the better message here is "Queue reset threshold is expected to elapse after waiting"

File: agent/stats/queue_test.go [hunk 26]
```diff
@@ -309,3 +310,55 @@ func TestCpuStatsSetNotSetToInfinity(t *testing.T) {
...
+		3649536,
+		3649536,
+		3649536,
+	}
+	// create a queue
+	queueLength := 3
+	queue := NewQueue(queueLength)
+
+	enoughDataPoints := queue.enoughDatapointsInBuffer()
+	assert.False(t, enoughDataPoints, "Queue is expected NOT having enough data points right after creation")
```
Comment (aaithal) [inline agent/stats/queue_test.go]: Grammar: "Queue is expected to not have enough data ..."

File: agent/stats/queue_test.go [hunk 27]
```diff
@@ -309,3 +310,55 @@ func TestCpuStatsSetNotSetToInfinity(t *testing.T) {
...
+	for i, time := range timestamps {
+		queue.Add(&ContainerStats{cpuUsage: cpuTimes[i], memoryUsage: memoryUtilizationInBytes[i], timestamp: time})
+	}
+
+	enoughDataPoints = queue.enoughDatapointsInBuffer()
+	assert.True(t, enoughDataPoints, "Queue is expected to have enough data points when it has more than 2 msgs queued")
+
+	queue.Reset()
+	enoughDataPoints = queue.enoughDatapointsInBuffer()
+	assert.False(t, enoughDataPoints, "Queue is expected NOT having enough data points right after RESET")
```
Comment (aaithal) [inline agent/stats/queue_test.go]: Grammar: "Queue is expected to not have enough data ..."

File: agent/acs/handler/acs_handler_test.go [hunk 28]
```diff
@@ -518,10 +518,14 @@ func TestHandlerReconnectDelayForInactiveInstanceError(t *testing.T) {
...
+			reconnectDelayTime := time.Now()
 			t.Logf("Delay between successive connections: %v", reconnectDelay)
-			if reconnectDelay < inactiveInstanceReconnectDelay {
-				t.Errorf("Reconnect delay %v is not less than inactive instance reconnect delay %v", reconnectDelay, inactiveInstanceReconnectDelay)
-			}
+			timeSubFuncSlopAllowed := 2 * time.Millisecond
+			// On windows platform, we found issue with time.Now().Sub(...) reporting 199.9989 even
+			// after the code has already waited for time.NewTimer(200)ms.
+			assert.WithinDuration(t, reconnectDelayTime, firstConnectionAttemptTime.Add(inactiveInstanceReconnectDelay), timeSubFuncSlopAllowed,
+				"Reconnect delay %v is expected within %v of inactive instance reconnect delay %v",
```
Comment (samuelkarp) [inline agent/acs/handler/acs_handler_test.go]: You don't need a message this detailed; the `assert` library will do this for you.

Comment (liwenwu-amazon) [inline agent/acs/handler/acs_handler_test.go]: You mean
1) I should remove comment (line 524 to line 525)?
or
2) rewording remove line 527?
thanks

Comment (samuelkarp) [inline agent/acs/handler/acs_handler_test.go]: Reword line 527 and remove the format string.  To see what I mean, you can temporarily make the assertion fail on your local machine and observe the error message that gets printed out.

**Comment by aaithal** [comment]
@liwenwu-amazon I think the code changes look good. Can you manually test if this fix has any side-effects on a long-running (`1h`) agent? We basically want to verify that:
1. Disconnect/Reconnect to ACS and TACS keep happening on a long running ECS Agent
2. Service/Cluster stats have a consistent value for a fixed workload on a long running ECS Agent

**Comment by liwenwu-amazon** [comment]
I have this image running for past 6 days.  So far, I have not seen any warning messages. Nor I have seen any abnormal stats on "Metrics" dashboard on AWS console.

The CPUUtilization is kept at 0.03% average and few spikes maxed @ 0.2% .  The CPU Reservation is stabilized at 50%. The MemoryUtilization is stabilized at 6% and memoryreservation is stabilized at 27%
